SCLOPF intermittently allows post-contingency line overloads (old "flaky" test issue)

### Version Checks (indicate both or one)

- [ ] I have confirmed this bug exists on the lastest [release](https://github.com/pypsa/pypsa/releases) of PyPSA.

- [x] I have confirmed this bug exists on the current [`master`](https://github.com/pypsa/pypsa/tree/master) branch of PyPSA.


### Issue Description

See `test_sclopf_scigrid.py` and SCLOPF code in `abstract.py`

Security‑constrained optimization on the SciGrid-DE example (with two line outages) sometimes yields LPF contingency flows exceeding thermal limits by up to 7% (e.g. Line 673: 1.073 p.u.). 

> FAILED test/test_sclopf_scigrid.py::test_optimize_security_constrained - AssertionError: 
> Arrays are not almost equal to 4 decimals
> Mismatched elements: 2 / 3 (66.7%)
> Max absolute difference among violations: 0.07296842
> Max relative difference among violations: 0.07296842
>  ACTUAL: array([1.    , 1.0246, 1.073 ])
>  DESIRED: array([1., 1., 1.])

Failures are intermittent and appear irregularly. Rerunning it locally with ubuntu & python 3.13 yields 1 in 30 fails, always with the same overload values.

The big problem here is that it breaks CI, also on master of PyPSA.

Reproduce: `pytest -q test/test_sclopf_scigrid.py::test_optimize_security_constrained` long enough or run the bash script below with `bash test.py N` for N runs.

A degeneracy of solution mostly likely plays a role here, but after some checks I *think* it is that degeneracy highlights there is a problem in core code of `abstract.py`, rather than degeneracy causes the problem.

It requires a proper look. Maybe @coroa @FabianHofmann 


```
#!/bin/bash

# Check if number of runs is provided
if [ $# -eq 0 ]; then
    echo "Usage: $0 <number_of_runs>"
    exit 1
fi

RUNS=$1
PASSED=0
FAILED=0

echo "Running test $RUNS times..."
echo "================================"

for i in $(seq 1 $RUNS); do
    echo -n "Run $i/$RUNS: "
    
    if uv run pytest test/test_sclopf_scigrid.py::test_optimize_security_constrained -q 2>/dev/null; then
        echo "✓ PASSED"
        ((PASSED++))
    else
        echo "✗ FAILED"
        ((FAILED++))
    fi
done

echo "================================"
echo "Results:"
echo "  Passed: $PASSED/$RUNS"
echo "  Failed: $FAILED/$RUNS"
echo "  Success rate: $(echo "scale=2; $PASSED * 100 / $RUNS" | bc)%"
```

### Reproducible Example

```python
pytest -q test/test_sclopf_scigrid.py::test_optimize_security_constrained
```

### Expected Behavior

passed tests

### Installed Versions

<details>
Replace this line.
</details>

--- a/docs/release-notes.md
+++ b/docs/release-notes.md
@@ -16,6 +16,13 @@ SPDX-License-Identifier: CC-BY-4.0
     to install the `master` branch, e.g. `pip install git+https://github.com/pypsa/pypsa`.
 -->
 
+## Upcoming Release
+
+### Bug Fixes
+
+- Fix [`n.graph()`][pypsa.network.graph.NetworkGraphMixin.graph] building edges in a non-deterministic order, which could make results that depend on the network's cycles differ between runs. In particular, security-constrained optimization (SCLOPF) now returns consistent results. (<!-- md:pr 1764 -->)
+
+
 ## [**v1.2.4**](https://github.com/PyPSA/PyPSA/releases/tag/v1.2.4) <small>27th June 2026</small> { id="v1.2.4" }
 
 ### Bug Fixes

--- a/docs/release-notes.md
+++ b/docs/release-notes.md
@@ -6,7 +6,6 @@ SPDX-License-Identifier: CC-BY-4.0
 
 # Release Notes
 
-<!--
 ## Upcoming Release
 
 !!! info "Upcoming Release"
@@ -14,9 +13,6 @@ SPDX-License-Identifier: CC-BY-4.0
     The features listed below have not yet been released, but will be included in the
     next update! If you would like to use these features in the meantime, you will need
     to install the `master` branch, e.g. `pip install git+https://github.com/pypsa/pypsa`.
--->
-
-## Upcoming Release
 
 ### Bug Fixes
 

--- a/pypsa/network/graph.py
+++ b/pypsa/network/graph.py
@@ -83,6 +83,8 @@ class NetworkGraphMixin:
         else:
             msg = "graph must be called with a Network or a SubNetwork"
             raise TypeError(msg)
+        # sort for a hash-seed-independent edge order (and cycle basis); see GH #1356
+        branch_components = sorted(branch_components)
 
         buses_i = n.c.buses.static.index
 

--- a/test/test_sclopf_scigrid.py
+++ b/test/test_sclopf_scigrid.py
@@ -2,6 +2,7 @@
 #
 # SPDX-License-Identifier: MIT
 
+import numpy as np
 import pandas as pd
 from numpy.testing import assert_almost_equal as equal
 
@@ -41,17 +42,15 @@ def test_optimize_security_constrained(scipy_network):
     n.c.generators.dynamic.p_set = n.c.generators.dynamic.p.copy()
     n.c.storage_units.dynamic.p_set = n.c.storage_units.dynamic.p.copy()
 
-    # TODO see https://github.com/PyPSA/PyPSA/issues/1356
+    # Check no lines are overloaded with the linear contingency analysis
+    p0_test = n.lpf_contingency(n.snapshots[0], branch_outages=branch_outages)
 
-    # # Check no lines are overloaded with the linear contingency analysis
-    # p0_test = n.lpf_contingency(n.snapshots[0], branch_outages=branch_outages)
+    # Check loading as per unit of s_nom in each contingency
+    max_loading = (
+        abs(p0_test.divide(n.passive_branches().s_nom, axis=0)).describe().loc["max"]
+    )
 
-    # # Check loading as per unit of s_nom in each contingency
-    # max_loading = (
-    #     abs(p0_test.divide(n.passive_branches().s_nom, axis=0)).describe().loc["max"]
-    # )
-
-    # arr_equal(max_loading, np.ones(len(max_loading)), decimal=4)
+    equal(max_loading, np.ones(len(max_loading)), decimal=4)
     equal(n.objective, 339758.4578, decimal=1)
 
     # === Dual variable assignment checks ===

